src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs

```diff
@@ -825,10 +825,108 @@ public bool IsPrefix(ReadOnlySpan<char> source, ReadOnlySpan<char> prefix, Compa
+        /// the match length information.
+        /// </remarks>
+        public unsafe bool IsPrefix(ReadOnlySpan<char> source, ReadOnlySpan<char> prefix, CompareOptions options, out int matchLength)
+        {
+            // Just like the existing IsPrefix overloads, we'll allow an early
+            // exit here before 'options' is validated.
+
+            if (prefix.IsEmpty)
+            {
+                matchLength = 0;
+                return true;
+            }
+
+            if ((options & ValidIndexMaskOffFlags) == 0)
+            {
+                // Common case: caller is attempting to perform a linguistic search.
+                // Pass the flags down to NLS or ICU unless we're running in invariant
+                // mode, at which point we normalize the flags to Orginal[IgnoreCase].
+
+                if (!GlobalizationMode.Invariant)
+                {
+                    int tempMatchLength;
+                    if (StartsWithCore(source, prefix, options, &tempMatchLength))
+                    {
+                        Debug.Assert(tempMatchLength >= 0 && tempMatchLength <= source.Length);
+                        matchLength = tempMatchLength;
+                        return true;
+                    }
+                    else
+                    {
+                        matchLength = default;
+                        return false;
+                    }
+                }
+                else if ((options & CompareOptions.IgnoreCase) == 0)
+                {
+                    goto ReturnOrdinal;
+                }
+                else
+                {
+                    goto ReturnOrdinalIgnoreCase;
+                }
+            }
+            else
+            {
+                // Less common case: caller is attempting to perform non-linguistic comparison,
+                // or an invalid combination of flags was supplied.
+
+                if (options == CompareOptions.Ordinal)
+                {
+                    goto ReturnOrdinal;
+                }
+                else if (options == CompareOptions.OrdinalIgnoreCase)
+                {
+                    goto ReturnOrdinalIgnoreCase;
+                }
+                else
+                {
+                    ThrowCompareOptionsCheckFailed(options);
+                }
+            }
+
+        ReturnOrdinal:
+            bool retVal = source.StartsWith(prefix);
+            goto OrdinalReturn;
+
+        ReturnOrdinalIgnoreCase:
+            retVal = source.StartsWithOrdinalIgnoreCase(prefix);
+
+        OrdinalReturn:
+            // Both Ordinal and OrdinalIgnoreCase match by individual code points in a non-linguistic manner.
+            // Non-BMP code points will never match BMP code points, so given UTF-16 inputs the match length
+            // will always be equivalent to the target string length.
+
+            matchLength = (retVal) ? prefix.Length : 0;
+            return retVal;
+        }
```

Comment:
this method is quite big and contains a lot of ifs and goto. Could it be simplified to something like this:

```cs
public unsafe bool IsPrefix(ReadOnlySpan<char> source, ReadOnlySpan<char> prefix, CompareOptions options, out int matchLength)
{
    if (prefix.IsEmpty)
    {
        matchLength = 0;
        return true;
    }
    
    int tempMatchLength = -1;
    bool result = (options & ValidIndexMaskOffFlags) == 0 && !GlobalizationMode.Invariant
        ? StartsWithCore(source, prefix, options, &tempMatchLength)
        : IsPrefix(source, prefix, options);
        
    matchLength = result
        ? (tempMatchLength >= 0 : tempMatchLength : prefix.Length)
        : 0;
}
```

Reply:
I bet we could do something like that. Will try it out shortly.